src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -812,101 +731,204 @@ public int getCurrentRunningCount() {
+     * starts it will attempt to start taking tasks, no further action is needed.
      */
-    protected Worker makeNewWorker() {
+    protected void makeNewWorker() {
       Worker w = new Worker(this, threadFactory);
-      currentPoolSize++;
       w.start();
-      
-      // will be added to available workers when done with first task
-      return w;
     }
     
     /**
-     * Shuts down the worker and ensures this now dead worker wont be used.
+     * Adds a worker to the head of the idle worker chain.
      * 
-     * @param w worker to shutdown
+     * @param worker Worker that is ready to become idle
      */
-    protected void killWorker(Worker w) {
-      // IMPORTANT** if this lock is removed, it is important to read the comment bellow
-      synchronized (workersLock) {
-        /* this will throw an exception if the worker has already stopped, 
-         * we want to ensure the pool size is not decremented more than once for a given worker.
-         * 
-         * We are able to stop first, because we are holding the workers lock.  In the future if we 
-         * try to reduce locking around here, we need to ensure that the worker is removed from the 
-         * available workers BEFORE stopping.
-         */
-        w.stopIfRunning();
-        currentPoolSize--;
-        // it may not always be here, but it sometimes can (for example when a worker is interrupted)
-        availableWorkers.remove(w);
+    protected void addWorkerToIdleChain(Worker worker) {
+      while (true) {
+        Worker casWorker = idleWorker.get();
+        // we can freely set this value until we get into the idle linked queue
+        worker.nextIdleWorker = casWorker;
+        if (idleWorker.compareAndSet(casWorker, worker)) {
+          break;
+        }
       }
     }
-    
+
     /**
-     * Called by the worker after it completes a task.  This is so that we can run any after task 
-     * cleanup, and make sure that the worker is now available for future tasks.
+     * Invoked when a worker becomes idle.  This will provide another task for that worker, or 
+     * block until a task is either ready, or the worker should be shutdown (either because pool 
+     * was shut down, or max pool size changed).
      * 
-     * @param worker worker that is now idle and ready for more tasks
+     * @param worker Worker which is now idle and ready for a task
+     * @return Task that is ready for immediate execution
      */
-    protected void workerDone(Worker worker) {
-      synchronized (workersLock) {
-        if (shutdownFinishing || currentPoolSize > maxPoolSize) {
-          killWorker(worker);
-        } else {
-          // always add to the front so older workers are at the back
-          availableWorkers.addFirst(worker);
-          
-          if (waitingForWorker) {
-            workersLock.notify();
+    public TaskWrapper workerIdle(Worker worker) {
```

Comment:
Here is the heart of the logic...as with many of these non-locking designs we loop for thread safe consistency.   This is invoked when a thread becomes idle and wants to get more work.  Here is the run down of what is happening here:

Before a thread ever parks it MUST queue itself first then recheck before it parks.  This is necessary because the state may have changed.  Once we queue, if the state changed then an unpark would have been issued and we would wake up immediately.  We try not to queue if we don't have to so we can maintain maximum throughput.  Of course the cost of that is having to double our work before we actually do park.

* Task is idle, we check if the pool is shutting down, or over sized, and thus the thread should be killed
* Otherwise we get a task
* If there is no task, we block forever till a task comes in
* If we got a task, we now must store an execute reference to ensure a task is not consumed multiple times (multiple threads likely got the same task)
* We check the delay, if it is not ready to run we must sleep.  If another thread has beat us and is doing a timed wait (which must be for the same task), then we wait forever, letting that thread handle the task without contention
* Assuming the task is ready for execution we verify that this thread can execute the task in a thread safe way (using a combination of the execute reference gotten earlier, as well as the queue state)  If another thread fails this check, then it will try to grab another task and retry it all again
* Weather we got a task, or not (maybe the pool shutdown for example)
* At the end, if we queued ourselves, it is now time to remove ourselves from the idle queue.  We must then wake up the next thread to ensure at least one thread is taking the active lead role (timed wait if necessary vs waiting forever)